### PR TITLE
[WIP]: Avatar size improvement

### PR DESCRIPTION
**Pull request by vvitto**
Closes [17738](https://github.com/ant-design/ant-design/issues/17738)

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Perfermance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Avatar component expansion `size` prop |
| 🇨🇳 Chinese | 阿凡达组件扩展`size'道具 |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

-----
[View rendered components/avatar/demo/grid.md](https://github.com/vvitto/ant-design/blob/avatar-size-improvement/components/avatar/demo/grid.md)
[View rendered components/avatar/index.en-US.md](https://github.com/vvitto/ant-design/blob/avatar-size-improvement/components/avatar/index.en-US.md)

### CODE REVIEW

**Comment by yoyo837** [comment]
混入了其他提交

**Comment by vvitto** [comment]
I would be grateful if someone would help me with translations.

File: components/avatar/__tests__/Avatar.test.js [hunk 1]
```diff
@@ -148,6 +148,11 @@ describe('Avatar Render', () => {
     expect(wrapper).toMatchRenderedSnapshot();
   });
 
+  it('support size is object', () => {
```
Comment (zombieJ) [inline components/avatar/__tests__/Avatar.test.js:151]: Should add screen diff size to test different snapshot

File: components/avatar/demo/grid.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+---
```
Comment (zombieJ) [inline components/avatar/demo/grid.md:1]: `grid.md` -> `responsive.md`

File: components/avatar/demo/grid.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+order: 5
+title:
+  zh-CN: Grid
+  en-US: Grid
+---
+
+## en-US
+
+Avatar size can be changed depends on screen size.
```
Comment (zombieJ) [inline components/avatar/demo/grid.md:10]: Chinese version:
> Avatar 支持响应式图标大小。

File: components/avatar/demo/grid.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+```tsx
+import React, { useState } from 'react';
+import { Avatar, Button } from 'antd';
+import { UserOutlined } from '@ant-design/icons';
+
+const sizes = { xs: 20, sm: 40, md: 60, lg: 80, xl: 100, xxl: 120 };
+
+const Autoset: React.FC = () => {
+  return (
+    <>
```
Comment (zombieJ) [inline components/avatar/demo/grid.md:21]: No need Fragment.

File: components/avatar/demo/grid.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+## en-US
+
+Avatar size can be changed depends on screen size.
+
+```tsx
+import React, { useState } from 'react';
+import { Avatar, Button } from 'antd';
+import { UserOutlined } from '@ant-design/icons';
+
+const sizes = { xs: 20, sm: 40, md: 60, lg: 80, xl: 100, xxl: 120 };
```
Comment (zombieJ) [inline components/avatar/demo/grid.md:17]: Just move `sizes` into component prop

File: components/avatar/index.en-US.md [hunk 6]
```diff
@@ -13,7 +13,7 @@ Avatars can be used to represent people or objects. It supports images, `Icon`s,
 | --- | --- | --- | --- | --- |
 | icon | custom icon type for an icon avatar | ReactNode | - |  |
 | shape | the shape of avatar | `circle` \| `square` | `circle` |  |
-| size | the size of the avatar | number \| `large` \| `small` \| `default` | `default` |  |
+| size | the size of the avatar | number \| `large` \| `small` \| `default` \| `{ xs: number, sm: number, ...}` | `default` |
```
Comment (zombieJ) [inline components/avatar/index.en-US.md:16]: Add version info about which version start to support responsive.

File: components/avatar/index.tsx [hunk 7]
```diff
@@ -102,11 +130,22 @@ export default class Avatar extends React.Component<AvatarProps, AvatarState> {
     }
   };
 
+  get size(): AvatarProps['size'] {
```
Comment (zombieJ) [inline components/avatar/index.tsx:133]: We do not use `set` | `get` in antd component. Let's just use `getSize` as function instead.

File: components/avatar/index.tsx [hunk 8]
```diff
@@ -102,11 +130,22 @@ export default class Avatar extends React.Component<AvatarProps, AvatarState> {
...
+    const { size } = this.props;
+
+    if (typeof size !== 'object') return size;
+
+    const screen = responsiveArray.find(breakpoint => this.state.screens[breakpoint]);
+
+    if (!screen) return Avatar.defaultProps.size;
```
Comment (zombieJ) [inline components/avatar/index.tsx:140]: Not use `defaultProps`

**Comment by afc163** [comment]
conflicted

**Comment by afc163** [comment]
ping

**Comment by vvitto** [comment]
Is it ok to use https://www.npmjs.com/package/css-mediaquery for proper matching of the screen size in matchMedia function?

**Comment by 07akioni** [comment]
> Is it ok to use https://www.npmjs.com/package/css-mediaquery for proper matching of the screen size in matchMedia function?

you may implement it like Col or Row

**Pull request closed without merging**